Idiamond/prometheus and statsd

idiamond-stripe opened the pull request:
This change replaces `go-metrics` with the standard Prometheus library and emits StatsD timing metrics for each request, rather than aggregating it in memory. 

This PR creates `metrics.go` files on the package level for each package it monitors and then emits metrics in the centralized `prometheus` package. This also creates a `statsd` package that can emit metrics to a StatsD sink on demand.

Avoiding in-memory metric aggregation means aggregation can be performed in the metrics engine like SIgnalFX or DataDog or Graphite. This allows users to more easily monitor the p99s as they'd like and determine when individual requests endure latency spikes, rather than simply seeing the view in aggregate.

- Replace go-metrics Prometheus metrics with the Prometheus go client library and emit all previous `go-metrics` metrics that way
- Replace go-metrics StatsD with a client that does not perform in-memory aggregation and emit timing metrics that way
- Add additional logging as to the duration of requests

Resolves #35

pingles commented:
@idiamond-stripe wow, thanks- this looks great.

I'm going to focus on trying to get #109 merged and then review this with a little more time. It all looks good- only thing I'd like to check is the names of metrics that are published/tracked.

Is there any chance you could add a `doc/telemetry.md` file that has a table of the different metric names and what they measure please? Would be super useful- both for this and afterwards for people operating kiam.

pingles commented:
@sambooo is there any chance this is something you could help go through and update to bring in line with the new handler and testing stuff we've done please?

idiamond-stripe commented:
@pingles, sorry, I haven't had a chance to work on this more. I'll set
aside some time this week to try and rebase it on top of your changes :)
On Mon, Jul 16, 2018 at 8:50 AM Paul Ingles <notifications@github.com>
wrote:

> @sambooo <https://github.com/sambooo> is there any chance this is
> something you could help go through and update to bring in line with the
> new handler and testing stuff we've done please?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/uswitch/kiam/pull/113#issuecomment-405293753>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AkAmU1qs5cfygzVpONZucMeyxsswPYizks5uHLY4gaJpZM4VGN7c>
> .
>

pingles commented:
That'd be awesome, thanks @idiamond-stripe. 

@sambooo hold off on picking this up for now then please

idiamond-stripe commented:
Alright @pingles, rebased on top of your changes and went ahead and added a `docs/metrics.md` as you suggested. :) 

That wasn't too bad, and hopefully everything here makes sense. Wasn't quite sure how to discuss the RPC metrics in the documentation, and if you have any insight that would be great!

idiamond-stripe commented on pkg/aws/metadata/handler_credentials.go:
```diff
@@ -36,9 +37,9 @@ type credentialsHandler struct {
-	credentialTimings := metrics.GetOrRegisterTimer("credentialsHandler", metrics.DefaultRegistry)
-	startTime := time.Now()
-	defer credentialTimings.UpdateSince(startTime)
+	timer := prometheus.NewTimer(handlerTimer.WithLabelValues("credentials"))
```
It's fine to just create a timer inline. A `timer` in the Prometheus library is just a single timing for a function call, so it makes sense to `defer` the completion of it until the end of the fn.

pingles commented on docs/metrics.md:
```diff
@@ -0,0 +1,59 @@
+### Prometheus
+#### Metadata Subsystem
+- `handler_latency_milliseconds` - Bucketed histogram of handler timings. Tagged by handler
+- `credential_fetch_error` - Number of errors fetching the credentials for a pod
```
Should this be renamed to `credential_fetch_errors_total` to indicate its an accumulating counter?

pingles commented on docs/metrics.md:
```diff
@@ -0,0 +1,59 @@
+#### Metadata Subsystem
+- `handler_latency_milliseconds` - Bucketed histogram of handler timings. Tagged by handler
+- `credential_fetch_error` - Number of errors fetching the credentials for a pod
+- `credential_encode_error` - Number of errors encoding credentials for a pod
```
Same as above, should this be `credential_encode_errors_total`? Or is it not an accumulating counter?

pingles commented on pkg/aws/sts/metrics.go:
```diff
@@ -0,0 +1,61 @@
+		prometheus.CounterOpts{
+			Namespace: "kiam",
+			Subsystem: "sts",
+			Name:      "error_issuing_count",
```
To match the other metrics could we rename to `issuing_errors` or `issuing_errors_total` please?

pingles commented on docs/metrics.md:
```diff
@@ -0,0 +1,59 @@
+- `cache_miss_total` - Number of cache misses to the metadata cache
+- `error_issuing_count` - Number of errors issuing credentials
+- `assumerole_timing_milliseconds` - Bucketed histogram of assumeRole timings
+- `assume_role_executing_total` - Number of assume role calls currently executing
```
I'm not sure whether we should use the `_total` suffix for a gauge? Maybe rename to `assumerole_current` or `assumerole_current_operations`?

pingles commented:
I'm going to close this PR in favour of #131 which picks up your contribution with a few changes. Hope that's ok. Thanks again for all your hard work @idiamond-stripe!
